ee/query-service/anomaly/params.go

```diff
@@ -153,30 +161,76 @@ func prepareAnomalyQueryParams(req *v3.QueryRangeParamsV3, seasonality Seasonali
 	var pastGrowthPeriodStart, pastGrowthPeriodEnd int64
 	switch seasonality {
 	case SeasonalityWeekly:
-		pastGrowthPeriodStart = start - 14*24*time.Hour.Milliseconds()
-		pastGrowthPeriodEnd = start - 7*24*time.Hour.Milliseconds()
+		pastGrowthPeriodStart = start - 2*oneWeekOffset
+		pastGrowthPeriodEnd = start - 1*oneWeekOffset
 	case SeasonalityDaily:
-		pastGrowthPeriodStart = start - 2*time.Hour.Milliseconds()
-		pastGrowthPeriodEnd = start - 1*time.Hour.Milliseconds()
+		pastGrowthPeriodStart = start - 2*oneDayOffset
+		pastGrowthPeriodEnd = start - 1*oneDayOffset
 	case SeasonalityHourly:
-		pastGrowthPeriodStart = start - 2*time.Hour.Milliseconds()
-		pastGrowthPeriodEnd = start - 1*time.Hour.Milliseconds()
+		pastGrowthPeriodStart = start - 2*oneHourOffset
+		pastGrowthPeriodEnd = start - 1*oneHourOffset
 	}
 
 	pastGrowthQuery := &v3.QueryRangeParamsV3{
 		Start:          pastGrowthPeriodStart,
 		End:            pastGrowthPeriodEnd,
-		CompositeQuery: copyCompositeQuery(req),
+		CompositeQuery: req.CompositeQuery.Clone(),
 		Variables:      make(map[string]interface{}, 0),
 		NoCache:        false,
 	}
 	updateStepInterval(pastGrowthQuery)
 
+	var past2GrowthPeriodStart, past2GrowthPeriodEnd int64
+	switch seasonality {
+	case SeasonalityWeekly:
+		past2GrowthPeriodStart = start - 3*oneWeekOffset
+		past2GrowthPeriodEnd = start - 2*oneWeekOffset
+	case SeasonalityDaily:
+		past2GrowthPeriodStart = start - 3*oneDayOffset
+		past2GrowthPeriodEnd = start - 2*oneDayOffset
+	case SeasonalityHourly:
+		past2GrowthPeriodStart = start - 3*oneHourOffset
+		past2GrowthPeriodEnd = start - 2*oneHourOffset
+	}
+
+	past2GrowthQuery := &v3.QueryRangeParamsV3{
+		Start:          past2GrowthPeriodStart,
+		End:            past2GrowthPeriodEnd,
+		CompositeQuery: req.CompositeQuery.Clone(),
+		Variables:      make(map[string]interface{}, 0),
+		NoCache:        false,
+	}
+	updateStepInterval(past2GrowthQuery)
+
+	var past3GrowthPeriodStart, past3GrowthPeriodEnd int64
+	switch seasonality {
+	case SeasonalityWeekly:
```

Comment:
all of this can be moved to a helper which takes past as an argument ?

Reply:
I am not sure I follow. Please share what it would look like. I optimize for the code readability over lines of code. I believe it's easier for eyes to scan through as it is structured currently.

Reply:
I meant something like this

```
func getPastGrowthPeriod(seasonality string, start , past int64) {
	var growthStart, growthEnd int64
	switch seasonality {
	case SeasonalityWeekly:
		growthStart = start - past*oneWeekOffset
		growthEnd = start - (past-1)*oneWeekOffset
	case SeasonalityDaily:
		growthStart = start - past*oneDayOffset
		growthEnd = start - (past-1)*oneDayOffset
	case SeasonalityHourly:
		growthStart = start - past*oneHourOffset
		growthEnd = start - (past-1)*oneHourOffset
	}
	return growthStart, growthEnd
}
```

Though I don't have strong opinion, both are fine.

Reply:
I would prefer to keep it as is because it doesn't change much except reduce some lines.